### ISSUE TITLE
4.0.0-beta.5: Storage.Adapters.FS and deleting/overwritting files

### ISSUE THREAD

**Issue by Allmektiger** [keystonejs/keystone#4245]
Hello.
I want to make avatar loading for users in local filesystem, but have a little headache with it...

`var myStorage = new keystone.Storage({
    adapter: keystone.Storage.Adapters.FS,
    fs: {
        path: 'public/images/users',
        publicPath: '/images/users',
        whenExists: 'overwrite'
    },
});`
...
`avatar: { type: Types.File, storage: myStorage },`

Firstly, loading in admin UI works fine, but after removing with "Remove File" link and saving file doesn't deleting from disk. Is this expected behaviour or issue?
I know that deleting avatars from admin panel is not necessary, but this may be actually for other content with file uploading/removing from admin UI.

Secondly, i need in possibility to reload avatars. Easiest way that i see is "**whenExists: 'overwrite'**" option. But it require the same naming of files for users. For example by email: **em@a.il.png**. Unfortunately, in current documentation i see:
> The generateFilename() method no longer receives a reference to the item being updated, we may introduce this in a future update.

Are there any alternative ways to naming file with one name for one user?

Thirdly, are there any callbacks for uploading process? For example, scaling and crop big image after saving. I know about cloudinary, but i really need in local storaging.

Fourthly, i want to implement type validation, like **allowedTypes** in old **LocalFile**. Is it possible?

I was really saddened by deprecating old LocalFile field type. Is there an easy way to restore this type in my local repository until new full versions of File type appear, without downgrading to keystonejs 0.3.17?

**Comment by CruelMoney** [comment]
I'm having some problems updating existing files as well. 
I try to overwrite the existing file when updating the file, but it seems that the "whenExists" attribute does not have any effect. 

My API update endpoint :
```
FileData.model.findById(req.params.id).exec(function(err, item) {
    if (err) return res.apiError('database error', err);
    if (!item) return res.apiError('not found');

    var data = (req.method == 'POST') ? req.body : req.query;

    item.getUpdateHandler(req).process(data, function(err) {

      if (err) return res.apiError('create error', err);
       
      item.url =  ('/uploads/files/'+item.file.filename)
      res.apiResponse(item);

    });
  });
```

And the model configuration:
```
var myStorage = new keystone.Storage({
    adapter: keystone.Storage.Adapters.FS,
    fs: {
        path: keystone.expandPath('../public/uploads/files'), // required; path where the files should be stored
        publicPath: '/uploads/files', // path where files will be served
        whenExists: 'overwrite',
        generateFilename: (file, attempt, cb)=> {
            cb(null, file.originalname)
        },
    },
});
```

I get `error: 'EEXIST: file already exists,`

**Comment by andrearaponi** [comment]
Hello guys, i've the same problem! But is a possible delete the file when i click delete the button in backend? Keystone delete only link on db, i need to delete the file :( 
 This is my model 

`var myStorage = new keystone.Storage({
  adapter: keystone.Storage.Adapters.FS,
  fs: {
    path: 'public/images/prod',
	publicPath: '/prod',
	generateFilename: (file, attempt, cb)=> {
            cb(null, file.originalname)
        },
  },
});`

**Comment by Kalcode** [comment]
I ran into this too. I created a function that gathers all the images stored and all the image filenames in each model. Compare the two and delete whatever isn't in the database anymore. 

I believe running this once a month or a week would be little overhead. 

I also do the same with the images uploaded to S3. 

Another note often when saving a file it will be left in the root of the project instead with the copy in the appropriate folder. So I also had to write a function on save to gather any images in the root and delete them. 

Hope that gives ya'll some ideas for a solution.

**Comment by kevinboudreault** [comment]
I'm having the same issue on Windows 7. Unlink doesn't return any errors but the file is still there. I can't upload a file with the same name as an existing one. @Kalcode , could you post your code you used to compare db with existing files? Thx

**Comment by Kalcode** [comment]
```javascript
const keystone = require('keystone')

function getImages(collection, resolve) {
  collection.model.find({}, { 'image.filename': 1, _id: 0 }, (err, images) => {
    if (err) console.log(err)
    const list = []
    images.forEach((obj) => {
      obj.image.filename && list.push(obj.image.filename)
    })
    resolve(list)
  })
}

const activityList = function() {
  return new Promise((resolve, reject) => {
    const StaffStory = keystone.list('Activity')
    getImages(StaffStory, resolve)
  })
}

const storyList = function() {
  return new Promise((resolve, reject) => {
    const Stroy = keystone.list('Story')
    getImages(Stroy, resolve)
  })
}

exports = module.exports = function() {
  return new Promise((resolve, reject) => {
    const collections = [storyList(), activityList()]
    Promise.all(collections)
    .then(list => {
      resolve({ database: list.reduce((acc, val) => acc.concat(val), []) })
    })
    .catch(err => console.log(err))
  })
}
```

I import the above as dbList and then i gather whats in the directory or s3. This is the directory code, then i add the two list to a new object and filter the list if its not in the database list. 

This leaves me with only the list thats not in the current database. 

Then I unlink them. 

```javascript
  var appDir = path.dirname(require.main.filename)
  var imageFolder = appDir + '/public/uploads/'

  const dirList = new Promise((resolve, reject) => {
    fs.readdir(imageFolder, (err, files) => {
      if (err) console.log(err)
      const list = []
      files.forEach(file => {
        list.push(file)
      })
      resolve({ directory: list })
    })
  })

  Promise.all([dirList, dbList()])
  .then((arrayOfLists) => {
    return new Promise((resolve, reject) => {
      const lists = {}
      arrayOfLists.forEach((item) => {
        Object.assign(lists, item)
      })

      const filtered = lists.directory.filter((filename) => {
        return lists.database.indexOf(filename) === -1
      })

      filtered.forEach((filename) => {
        const filePath = imageFolder + filename
        fs.unlink(filePath, err => console.log(err))
      })
      resolve(filtered)
    })
  })
  .then(filtered => res.json({ deleted: filtered }))
  .catch(err => console.log(err))
```

Probably not the most efficient code but I wrote it quickly to solve the issue and only run it once a week currently. But we don't get enough images backed-up that probably once a month would be enough.

**Comment by Noviny** [comment]
Resolved by #4466
